raft/node.go

```diff
@@ -49,43 +49,90 @@ func (a *SoftState) equal(b *SoftState) bool {
 // Ready encapsulates the entries and messages that are ready to read,
 // be saved to stable storage, committed or sent to other peers.
 // All fields in Ready are read-only.
+//
+// Handling the Ready can be done with or without performance optimizations.
+// Applications that don't need high throughput should eschew performance
+// optimizations for simplicity. These applications will use the following
+// sequence of steps:
+//
+// 1 apply snapshot (if any) and sync
+// 2 atomically append Entries (if any) and update HardState (if any).
+// 3. sync the Entries and HardState if MustSync is set.
+// 4. send messages
+// 5. apply CommittedEntries
+//
+// If for some reason it is impossible to append the Entries atomically with
+// updating the HardState, the entries (which are always synced) must be
+// appended and synced before the HardState is updated. This avoids the
+// situation in which the HardState references a Commit index which is not
+// durably appended yet. This does not penalize performance because updates to
+// the HardState must only be synced when MustSync is set, and those occur only
```

Comment:
This isn't actually true. Unfortunately, we set `MustSync` to true if there are entries to append but no change to `HardState.Vote` or `HardState.Term`. We might want to expose a new `MustSyncHardState` or something.

Reply:
Oh, right. I forgot that what I wanted to do was to change this to `MustSyncHardState` and add an accessor `func (rd *Ready) MustSync() { return len(rd.Entries) > 0 || rd.MustSyncHardState }`.